ksql-rest-app/src/test/java/io/confluent/ksql/rest/server/computation/StatementExecutorTest.java

```diff
@@ -412,14 +455,185 @@ public void shouldEnforceReferentialIntegrity() {
+    handleStatement(
+        statementExecutor, dropStreamCommand3, dropStreamCommandId3, Optional.empty());
 
     final Optional<CommandStatus> dropStreamCommandStatus3 =
         statementExecutor.getStatus(dropStreamCommandId3);
     assertThat(dropStreamCommandStatus3.get().getStatus(),
                CoreMatchers.equalTo(CommandStatus.Status.SUCCESS));
+  }
+
+  private Query mockCSASQuery(final String name) {
+    final Query mockQuery = mock(Query.class);
+    final QuerySpecification mockQuerySpec = mock(QuerySpecification.class);
+    final Table mockRelation = mock(Table.class);
+    expect(mockQuery.getQueryBody()).andStubReturn(mockQuerySpec);
+    expect(mockQuery.getLimit()).andStubReturn(Optional.empty());
+    expect(mockQuerySpec.getInto()).andStubReturn(mockRelation);
+    expect(mockRelation.getName()).andStubReturn(QualifiedName.of(name));
+    return mockQuery;
+  }
+
+  private CreateStreamAsSelect mockCSAS(final String name) {
+    final CreateStreamAsSelect mockStatement = mock(CreateStreamAsSelect.class);
+    expect(mockStatement.getName()).andStubReturn(QualifiedName.of(name));
+    expect(mockStatement.getQuery()).andStubReturn(mockCSASQuery(name));
+    expect(mockStatement.getProperties()).andStubReturn(Collections.emptyMap());
+    expect(mockStatement.getPartitionByColumn()).andStubReturn(Optional.empty());
+    return mockStatement;
+  }
+
+  private DropStream mockDropStream(final String name) {
+    final DropStream mockDropStream = mock(DropStream.class);
+    expect(mockDropStream.getName()).andStubReturn(QualifiedName.of(name));
+    expect(mockDropStream.getStreamName()).andStubReturn(QualifiedName.of(name));
+    return mockDropStream;
+  }
+
+  private Statement mockTerminate(final QueryId queryId) {
+    final TerminateQuery mockStatement = mock(TerminateQuery.class);
+    expect(mockStatement.getQueryId()).andStubReturn(queryId);
+    return mockStatement;
+  }
+
+  private PersistentQueryMetadata mockReplayCSAS(
+      final String statement,
+      final String name,
+      final QueryId queryId) {
+    final CreateStreamAsSelect mockCSAS = mockCSAS(name);
+    final PersistentQueryMetadata mockQuery = mock(PersistentQueryMetadata.class);
+    expect(mockQuery.getQueryId()).andStubReturn(queryId);
+    expect(mockParser.parseSingleStatement(statement)).andReturn(mockCSAS);
+    expect(
+        mockEngine.addInto(
+            anyObject(), anyObject(), anyObject(), anyObject(), anyObject(), anyBoolean()))
+        .andStubReturn(mockCSASQuery(name));
+    expect(mockEngine.getMetaStore()).andStubReturn(mockMetaStore);
+    expect(mockMetaStore.getSource(name)).andStubReturn(null);
+    expect(mockEngine.buildMultipleQueries(eq(statement), anyObject(), anyObject()))
+        .andReturn(Collections.singletonList(mockQuery));
+    return mockQuery;
+  }
+
+  @Test
+  public void shouldSkipStartForTerminatedQueriesWhenReplayingLog() {
+    // Given:
+    final String csas = "CSAS";
```

Comment:
IMHO, this would be better inlined to aid readability.  Or move to constants e.g. `CSAS_STATEMENT`...

Likewise, for `terminate` and `foo` and probably `queryId` too. They can all be constants or inlined.